client/rpc-spec-v2/src/chain_head/subscription.rs

```diff
@@ -20,266 +20,1045 @@
 use sp_blockchain::Error;
 use sp_runtime::traits::Block as BlockT;
 use std::{
-	collections::{hash_map::Entry, HashMap, HashSet},
+	collections::{hash_map::Entry, HashMap},
 	sync::Arc,
+	time::{Duration, Instant},
 };
 
 /// Subscription management error.
-#[derive(Debug)]
+#[derive(Debug, thiserror::Error)]
 pub enum SubscriptionManagementError {
 	/// The block cannot be pinned into memory because
 	/// the subscription has exceeded the maximum number
 	/// of blocks pinned.
+	#[error("Exceeded pinning limits")]
 	ExceededLimits,
 	/// Error originated from the blockchain (client or backend).
+	#[error("Blockchain error {0}")]
 	Blockchain(Error),
+	/// The database does not contain a block hash.
+	#[error("Block hash is absent")]
+	BlockHashAbsent,
 	/// The database does not contain a block header.
+	#[error("Block header is absent")]
 	BlockHeaderAbsent,
+	/// The specified subscription ID is not present.
+	#[error("Subscription is absent")]
+	SubscriptionAbsent,
 	/// Custom error.
+	#[error("Subscription error {0}")]
 	Custom(String),
 }
 
+// Blockchain error does not implement `PartialEq` needed for testing.
+impl PartialEq for SubscriptionManagementError {
+	fn eq(&self, other: &SubscriptionManagementError) -> bool {
+		match (self, other) {
+			(Self::ExceededLimits, Self::ExceededLimits) |
+			// Not needed for testing.
+			(Self::Blockchain(_), Self::Blockchain(_)) |
+			(Self::BlockHashAbsent, Self::BlockHashAbsent) |
+			(Self::SubscriptionAbsent, Self::SubscriptionAbsent) => true,
+			(Self::Custom(lhs), Self::Custom(rhs)) => lhs == rhs,
+			_ => false,
+		}
+	}
+}
+
 impl From<Error> for SubscriptionManagementError {
 	fn from(err: Error) -> Self {
 		SubscriptionManagementError::Blockchain(err)
 	}
 }
 
-/// Inner subscription data structure.
-struct SubscriptionInner<Block: BlockT> {
+/// The state machine of a block of a single subscription ID.
+///
+/// # Motivation
+///
+/// Each block is registered twice: once from the `BestBlock` event
+/// and once from the `Finalized` event.
+///
+/// The state of a block must be tracked until both events register the
+/// block and the user calls `unpin`.
+///
+/// Otherwise, the following race might happen:
+///  T0. BestBlock event: hash is tracked and pinned in backend.
+///  T1. User calls unpin: hash is untracked and unpinned in backend.
+///  T2. Finalized event: hash is tracked (no previous history) and pinned again.
+///
+/// # State Machine Transition
+///
+/// ```ignore
+///                   (register)
+/// [ REGISTERED ]  ---------------> [ FULLY REGISTERED ]
+///       |                              |
+///       | (unpin)                      | (unpin)
+///       |                              |
+///       V           (register)         V
+/// [ UNPINNED ]  -----------------> [ FULLY UNPINNED ]
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+enum BlockStateMachine {
+	/// The block was registered by one event (either `Finalized` or `BestBlock` event).
```

Comment:
Just to check my understanding, I thought only `BestBlock` would lead to the `Registered` state, and then `Finalized` leads to `FullyRegistered`  (assuming unpin hasn't been called in the meantime)?

So the names could also be, in my head:

```rust
BestBlock,
FinalizedBlock,
UnpinnedBestBlock,
UnpinnedFinalizedBlock
```

Or am I wrong in my understanding?